geomesa-feature/src/main/scala/org/locationtech/geomesa/feature/kryo/SimpleFeatureSerializer.scala

```diff
@@ -0,0 +1,346 @@
+    decodingsCache.get.get(sft.toString).flatMap(_.get) match {
+      case Some(functions) => functions
+      case None =>
+        val decodings = sft.getAttributeDescriptors.map { d =>
+          matchDecode(d.getType.getBinding, d.getUserData)
+        }.zipWithIndex
+        decodingsCache.get.put(sft.toString, new SoftReference(decodings))
+        decodings
+    }
+
+  /**
+   * Finds an decoding function based on the input type
+   *
+   * @param clas
+   * @param metadata
+   * @return
+   */
+  def matchDecode(clas: Class[_], metadata: jMap[Object, Object]): (Input) => AnyRef = clas match {
+
+    case c if classOf[String].isAssignableFrom(c) =>
+      (in: Input) => in.readString()
+
+    case c if classOf[java.lang.Integer].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else in.readInt().asInstanceOf[AnyRef]
+
+    case c if classOf[java.lang.Long].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else in.readLong().asInstanceOf[AnyRef]
+
+    case c if classOf[java.lang.Double].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else in.readDouble().asInstanceOf[AnyRef]
+
+    case c if classOf[java.lang.Float].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else in.readFloat().asInstanceOf[AnyRef]
+
+    case c if classOf[java.lang.Boolean].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else in.readBoolean().asInstanceOf[AnyRef]
+
+    case c if classOf[Date].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) null else new Date(in.readLong())
+
+    case c if classOf[UUID].isAssignableFrom(c) =>
+      (in: Input) => if (in.readByte() == NULL_BYTE) {
+        null
+      } else {
+        val mostSignificantBits = in.readLong()
+        val leastSignificantBits = in.readLong()
+        new UUID(mostSignificantBits, leastSignificantBits)
+      }
+
+    case c if classOf[Geometry].isAssignableFrom(c) =>
+      (in: Input) => {
+        val length = in.readInt(true)
+        if (length > 0) {
+          val bytes = new Array[Byte](length)
+          in.read(bytes)
+          WKBUtils.read(bytes)
+        } else {
+          null
+        }
+      }
+
+    case c if classOf[jList[_]].isAssignableFrom(c) =>
+      val subtype = metadata.get("subtype").asInstanceOf[Class[_]]
+      val subDecoding = matchDecode(subtype, null)
+
+      (in: Input) => {
+        val length = in.readInt()
+        if (length < 0) {
+          null
+        } else {
+          val list = new java.util.ArrayList[Object](length)
+          (0 until length).foreach(_ => list.add(subDecoding(in)))
+          list
+        }
+      }
+
+    case c if classOf[jMap[_, _]].isAssignableFrom(c) =>
+      val keyClass      = metadata.get("keyclass").asInstanceOf[Class[_]]
+      val valueClass    = metadata.get("valueclass").asInstanceOf[Class[_]]
```

Comment:
(picky) Factoring out the magic strings?  If you want to...  See also, AvroSimpleFeatureUtils and SimpleFeatureTypes.